components/resolved_ts/src/scanner.rs

```diff
@@ -82,109 +105,63 @@ impl<T: 'static + CdcHandle<E>, E: KvEngine> ScannerPool<T, E> {
         }
     }
 
-    pub fn spawn_task(&self, mut task: ScanTask) {
+    pub fn spawn_task(&self, mut task: ScanTask, concurrency_semaphore: Arc<Semaphore>) {
         let cdc_handle = self.cdc_handle.clone();
         let fut = async move {
             if let Some(backoff) = task.backoff {
                 RTS_INITIAL_SCAN_BACKOFF_DURATION_HISTOGRAM.observe(backoff.as_secs_f64());
-                if let Err(e) = GLOBAL_TIMER_HANDLE
+                let mut backoff = GLOBAL_TIMER_HANDLE
                     .delay(std::time::Instant::now() + backoff)
                     .compat()
-                    .await
-                {
-                    error!("failed to backoff"; "err" => ?e);
+                    .fuse();
+                futures::select! {
+                    res = backoff => if let Err(e) = res {
+                        error!("failed to backoff"; "err" => ?e);
+                    },
+                    _ = &mut task.cancelled => {}
                 }
-                if (task.is_cancelled)() {
+                if task.is_cancelled() {
                     return;
                 }
             }
+            let _permit = concurrency_semaphore.acquire().await;
```

Comment:
Maybe it would be better to have a metrics about pending task count which is used to see if the concurrency is enough.  Or if we already have pending task count metrics, then we can re-use that one.

Reply:
There is a metrics about tasks with labels "total", "finished" and "error", pending equals to `total - finished`.